Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm

```diff
@@ -0,0 +1,351 @@
+{
+    ASSERT(context);
+}
+
+WebExtensionCallbackHandler::WebExtensionCallbackHandler(JSContextRef context, JSObjectRef resolveFunction, JSObjectRef rejectFunction)
+    : m_callbackFunction(resolveFunction)
+    , m_rejectFunction(rejectFunction)
+    , m_globalContext(JSContextGetGlobalContext(context))
+{
+    ASSERT(context);
+    ASSERT(resolveFunction);
+    ASSERT(rejectFunction);
+
+    JSValueProtect(m_globalContext.get(), m_callbackFunction);
+    JSValueProtect(m_globalContext.get(), m_rejectFunction);
+}
+
+WebExtensionCallbackHandler::~WebExtensionCallbackHandler()
+{
+    if (m_callbackFunction)
+        JSValueUnprotect(m_globalContext.get(), m_callbackFunction);
+
+    if (m_rejectFunction)
+        JSValueUnprotect(m_globalContext.get(), m_rejectFunction);
+}
+
+JSValue *WebExtensionCallbackHandler::callbackFunction() const
+{
+    if (!m_globalContext || !m_callbackFunction)
+        return nil;
+
+    return [JSValue valueWithJSValueRef:m_callbackFunction inContext:[JSContext contextWithJSGlobalContextRef:m_globalContext.get()]];
+}
+
+template<size_t ArgumentCount>
+id callWithArguments(JSObjectRef callbackFunction, JSRetainPtr<JSGlobalContextRef>& globalContext, std::array<JSValueRef, ArgumentCount>&& arguments)
+{
+    if (!globalContext || !callbackFunction)
+        return nil;
+    return toNSObject(globalContext.get(), JSObjectCallAsFunction(globalContext.get(), callbackFunction, nullptr, ArgumentCount, arguments.data(), nullptr));
+}
+
+void WebExtensionCallbackHandler::reportError(NSString *message)
+{
+    if (!m_globalContext)
+        return;
+
+    if (m_runtime) {
+        m_runtime->reportErrorForCallbackHandler(*this, message, m_globalContext.get());
+        return;
+    }
+
+    if (!m_rejectFunction)
+        return;
+
+    JSValue *error = [JSValue valueWithNewErrorFromMessage:message inContext:[JSContext contextWithJSGlobalContextRef:m_globalContext.get()]];
+
+    callWithArguments<1>(m_rejectFunction, m_globalContext, {
+        toJSValue(m_globalContext.get(), error)
+    });
+}
+
+id WebExtensionCallbackHandler::call()
+{
+    return callWithArguments<0>(m_callbackFunction, m_globalContext, { });
+}
+
+id WebExtensionCallbackHandler::call(id argument)
+{
+    return callWithArguments<1>(m_callbackFunction, m_globalContext, {
+        toJSValue(m_globalContext.get(), argument)
+    });
+}
+
+id WebExtensionCallbackHandler::call(id argumentOne, id argumentTwo)
+{
+    return callWithArguments<2>(m_callbackFunction, m_globalContext, {
+        toJSValue(m_globalContext.get(), argumentOne),
+        toJSValue(m_globalContext.get(), argumentTwo)
+    });
+}
+
+id WebExtensionCallbackHandler::call(id argumentOne, id argumentTwo, id argumentThree)
+{
+    return callWithArguments<3>(m_callbackFunction, m_globalContext, {
+        toJSValue(m_globalContext.get(), argumentOne),
+        toJSValue(m_globalContext.get(), argumentTwo),
+        toJSValue(m_globalContext.get(), argumentThree)
+    });
```

Comment:
I'm a little surprised there aren't helper methods like this anywhere in the tree.

Reply:
It is the wild C++ west sometimes. But this is almost verbatim from the Safari copy.